feat: introduce `libp2p-identity` crate

thomaseizinger opened the pull request:
## Description

This patch combines the `libp2p_core::identity` and `libp2p_core::peer_id` modules into a new crate: `libp2p-identity`.

Resolves https://github.com/libp2p/rust-libp2p/issues/3349.

<!-- Please write a summary of your changes and why you made them.-->
<!-- This section will appear as the commit message after merging. Please craft it accordingly. -->

## Notes

This initial patch only moves the code and adds feature flags. What is yet undecided is how we want to adjust the public API as part of this. The goal would be that the first release (0.1) is backwards compatible with what we currently ship. In a 2nd release, we then harden the public API such that breaking changes will be very unlikely in the future. As part of `0.2`, I also want to remove the dependency on `multiaddr` to enable https://github.com/multiformats/rust-multiaddr/issues/73 which requires `multiaddr` to depend on `libp2p-identity`.

What do people think about this plan? @mxinden @jxs @elenaf9 

<!-- Any notes or remarks you'd like to make about the PR. -->

## Links to any relevant issues

<!-- Reference any related issues.-->

## Open Questions

<!-- Unresolved questions, if any. -->

## Further tasks

- [ ] Seal public API by making `Keypair` and `PublicKey` opaque structs. I think we should **not** have conditional enum variants. Those are a hazard. Someone could be exhaustively matching on them with one feature combination and adding a dependency that activates another suddenly causes the program to fail to compile. Even worse, this can happen in dependencies that don't add the necessary `cfg`s.

## Change checklist

<!-- Please add a Changelog entry in the appropriate crates and bump the crate versions if needed. See <https://github.com/libp2p/rust-libp2p/blob/master/docs/release.md#development-between-releases>-->

- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] A changelog entry has been made in the appropriate crates

thomaseizinger committed:
Introduce `libp2p-identity`

thomaseizinger committed:
Deprecate `PeerId::try_from_multiaddr`

thomaseizinger committed:
Add missing backtick

thomaseizinger committed:
Fix build errors

thomaseizinger committed:
Add package keys

thomaseizinger committed:
Move PeerId tests to `libp2p-identity`

thomaseizinger committed:
Move PeerId benches to `libp2p-identity`

thomaseizinger committed:
Migrate workspace to depend on libp2p-identity

thomaseizinger committed:
Add dummy keypair implementation

This allows crates to depend on the Keypair type without activing one
of the algorithms.

thomaseizinger committed:
Add missing re-exports of other key types

thomaseizinger committed:
Remove unnecessary docs

These docs will never render because this code is only available
without any features and we build docs with all features.

thomaseizinger committed:
Merge branch 'master' into feat/libp2p-identity

thomaseizinger committed:
Make `decoding_unsupported` unconditionally available

thomaseizinger committed:
Merge branch 'master' into feat/libp2p-identity

thomaseizinger committed:
Introduce ed25519 feature flag

thomaseizinger committed:
Sort stuff alphabetically

thomaseizinger committed:
Deprecate direct access to inner keypair and public key

thomaseizinger committed:
Use correct rsa feature flag

thomaseizinger committed:
Fix deprecated imports in tests

thomaseizinger committed:
Remove unused import

thomaseizinger committed:
Add deprecations for the feature-flags in libp2p-core

thomaseizinger committed:
Remove unnecessary feature gates

Now that we have dummy public keys, we no longer need this.

thomaseizinger committed:
More fixes

thomaseizinger committed:
Fix bad feature-gate

thomaseizinger committed:
More bad feature gates

thomaseizinger committed:
Merge branch 'master' into feat/libp2p-identity

thomaseizinger committed:
Merge branch 'master' into feat/libp2p-identity

thomaseizinger commented on core/Cargo.toml:
```diff
@@ -11,61 +11,44 @@ keywords = ["peer-to-peer", "libp2p", "networking"]
+secp256k1 = [ "libp2p-identity/secp256k1" ]
+ecdsa = [ "libp2p-identity/ecdsa" ]
+rsa = [ "libp2p-identity/rsa" ]
+ed25519 = [ "libp2p-identity/ed25519" ]
```
These are all deprecated and will be removed in the future.

mxinden replied:
Just to make sure we don't miss it, this is a breaking change. Previously `libp2p-core` would offer `ed25519` by default.

thomaseizinger replied:
Right, I would like it to not be a breaking change so I'll set this as a default feature, nice catch!

thomaseizinger replied:
I removed the feature from `libp2p-core` as it would be a breaking change either way because users could depend on it with `default-features = false`.

mxinden replied:
Sounds good to me. Thanks for the change.

thomaseizinger commented on identity/src/keypair.rs:
```diff
@@ -67,42 +55,100 @@ use std::convert::{TryFrom, TryInto};
 pub enum Keypair {
     /// An Ed25519 keypair.
+    #[cfg(feature = "ed25519")]
+    #[deprecated(
+        since = "0.1.0",
+        note = "This enum will be made opaque in the future, use `Keypair::into_ed25519` instead."
+    )]
```
The plan with this is that once users have migrated to the `into_` functions, we can make `Keypair` a struct and have the enum as a private field with `cfg`'d variants. That moves all the code that touches the conditional variants under our control where we can assure that it works under all feature-flag combinations. It will also allow us to remove the `keypair_dummy.rs` file again.

thomaseizinger committed:
Merge branch 'master' into feat/libp2p-identity

thomaseizinger committed:
Merge branch 'master' into feat/libp2p-identity

thomaseizinger committed:
Merge branch 'master' into feat/libp2p-identity

mxinden reviewed (commented):
I am sorry for the delay here.

I am in favor of the direction this is going.

Do I understand correctly that the main motivation for this change is to make `multiaddr`'s `/p2p` type safe by using `PeerId` from `libp2p-identity?

The proposals around hardening the API sound good to me in words. That said, I would need to see them in code and thus see their implications to really judge.

In case we go for the breaking change in `libp2p-core`, I am happy to push the version bumps right before merging.

thomaseizinger committed:
Merge branch 'master' into feat/libp2p-identity

thomaseizinger committed:
Don't feature flag ed25519 in libp2p-core

This would otherwise be a breaking change for users that depend on it with `default-features = false`.

thomaseizinger commented on identity/src/peer_id.rs:
```diff
@@ -94,6 +95,10 @@ impl PeerId {
+    #[deprecated(
+        since = "0.1.0",
+        note = "`Multiaddr` will become type-safe in a future release, making this function obsolete. See https://github.com/multiformats/rust-multiaddr/issues/73."
+    )]
```
Does it make sense to have this deprecation warning? At this stage, there isn't anything users can do about it because https://github.com/multiformats/rust-multiaddr/issues/73 isn't implemented yet. But at the same time, I want to tell users that this will go away. Thoughts?

mxinden replied:
I don't think it makes sense to advertise features through deprecation warnings before they are in place. I advocate for removing the deprecation warning.

thomaseizinger committed:
Fix deprecated usages

thomaseizinger commented on core/src/lib.rs:
```diff
@@ -55,25 +50,78 @@ mod peer_record_proto {
-mod peer_id;
+#[deprecated(since = "0.39.0", note = "Depend on `libp2p-identity` instead.")]
+pub mod identity {
+    pub use libp2p_identity::Keypair;
+    pub use libp2p_identity::PublicKey;
```
@obi1kenobi I think we might have hit another bug here! `cargo semver-checks` is reporting these two as missing. I've tried with type aliases as well and it doesn't make a difference which makes me think that I am doing something wrong here? It did work in my testing though, i.e. I got heaps of deprecation warnings but stuff was still working when I developed this feature.

obi1kenobi replied:
This is a move to another crate + re-export, right? Semver-checking right now only generates the tested crate's rustdoc JSON, so it doesn't see cross-crate re-exports: https://github.com/obi1kenobi/cargo-semver-checks/issues/355

It's fixable issue with a clear path forward. Just a matter of engineering cycles, which are always in short supply.

Thankfully, once you publish the next version of the affected crate, the false-positive will go away since the baseline will shift to that new version.

thomaseizinger committed:
Merge branch 'master' into feat/libp2p-identity

mxinden reviewed (approved):
Change looks good to me minus the deprecation warning.

thomaseizinger committed:
Merge branch 'master' into feat/libp2p-identity

thomaseizinger committed:
Properly generate proto bindings

thomaseizinger committed:
Adapt libp2p-identity to new proto structure

thomaseizinger committed:
Remove deprecation warning

thomaseizinger committed:
Generate proto files with correct compiler version

thomaseizinger committed:
Move changelog entry to correct version

thomaseizinger committed:
Amend changelog entry

thomaseizinger committed:
Merge branch 'master' into feat/libp2p-identity

thomaseizinger committed:
Merge branch 'master' into feat/libp2p-identity

thomaseizinger committed:
Fix bad merge

thomaseizinger committed:
Merge branch 'master' into feat/libp2p-identity

thomaseizinger committed:
Fix deprecation warnings and compile errors

thomaseizinger committed:
Generate proto bindings correctly

thomaseizinger committed:
Merge branch 'master' into feat/libp2p-identity

mxinden committed:
Merge branch 'master' into feat/libp2p-identity

mxinden committed:
Merge branch 'master' into feat/libp2p-identity
